### DISCUSSION TITLE
Issue with User Authentication in Kanidm-Unixd (v1.4)

### DISCUSSION THREAD

**Discussion by frateralexander** [kanidm/kanidm#3099] [General]
Hi,

I am encountering issues with user authentication using the kanidm-unixd (v1.4) service for POSIX users. After setting up Kanidm and configuring it to use PAM and NSS for user authentication, I am facing challenges allowing users to log in, and the logs suggest that there are issues related to group configuration in pam_allowed_login_groups. I’ve outlined my environment, steps taken, and the errors encountered below.

I would appreciate any guidance on how to resolve this issue, as I have followed the documentation but might be missing something critical in my configuration.

My environment:
- Kanidm version: 1.4
- Kanidm-unixd version: 1.4
- Operating system: Ubuntu 22.04
- Setup: Using kanidm-unixd for POSIX user authentication via PAM and NSS.

Steps taken:

Initial setup: In the configuration file /etc/kanidm/unixd, I left the pam_allowed_login_groups field empty to allow all POSIX groups to authenticate:

`pam_allowed_login_groups = [ ]`

Login attempt: When trying to authenticate using the command `su - usuario1`, I receive the following error in the logs:

`Oct 11 14:34:54 kanidm-server kanidm_unixd[3634]: Cannot authenticate users, no allowed groups in configuration!`

Attempt to add specific groups: I then tried to explicitly add the group usuario1 to the pam_allowed_login_groups field:

`pam_allowed_login_groups = ["usuario1"]
`

The error persisted:

`Oct 11 14:37:09 kanidm-server kanidm_unixd[3656]: Cannot authenticate users, no allowed groups in configuration!
`
Detailed logs: When checking the logs, I can see that the authentication request is reaching Kanidm, but the kanidm-unixd daemon is returning that no allowed groups were configured, even after explicitly adding the groups.

Logs from the login attempt with the user usuario1:

```
Oct 11 14:34:54 kanidm-server kanidm_unixd[3634]: Cannot authenticate users, no allowed groups in configuration!
Oct 11 14:37:09 kanidm-server kanidm_unixd[3656]: Cannot authenticate users, no allowed groups in configuration!
```
Additionally, the POSIX information appears to be correctly configured:

```
root@kanidm-server:/home/vagrant# kanidm person posix show usuario1
---
spn: usuario1@localhost
name: usuario1
displayname: Usuario Um    
shell: /bin/bash
group: [ spn: usuario1@localhost, gidnumber: 1003 name: usuario1 ]
```
And the group was also properly configured:

```
root@kanidm-server:/home/vagrant# kanidm group posix show usuario1
[ spn: usuario1@localhost, gidnumber: 1003 name: usuario1 ]
```

Could someone help ?

Regards

**Comment by yaleman** [comment]
> pam_allowed_login_groups = [ ]

Just to clarify, empty is not "all", you need to specify a list.

Can you try it with a user that's not the same name as the group? They could be stomping each other, because "user-name group" is ... a weird thing POSIX-land. 

ie: try "user: usuario1" group: "testgroup" or similar.

**Reply by frateralexander** [reply]
Thank you for the reply! However, in Linux/Unix environments, it's quite common for the user and group names to be the same, as seen in the concept of User Private Groups (UPGs). This approach works well within POSIX systems without apparent issues. However, I understand that in Kanidm, having identical names for users and groups might cause conflicts due to the way attributes like GID and SPN are managed. I appreciate the suggestion and will try to follow your advice to use distinct names for testing.

**Comment by Firstyear** [comment]
> Initial setup: In the configuration file /etc/kanidm/unixd, I left the pam_allowed_login_groups field empty to allow all POSIX groups to authenticate:
> 
> `pam_allowed_login_groups = [ ]`

`[]` means no one can login. We have a default-deny policy, not default-allow, to prevent accidental mistakes creating a security issue. If you want to allow all users, then set this to `pam_allowed_login_groups = ["idm_all_accounts"]` and enable posix attrs on idm_all_accounts.

**Reply by frateralexander** [reply]
Thank you for the helpful guidance! Unfortunately, after setting pam_allowed_login_groups = ["idm_all_accounts"], the issue persists. I understand that [] means no one can log in due to the default-deny policy, and that specifying a group like idm_all_accounts should allow access once POSIX attributes are enabled on the group. However, even after following this recommendation, I'm still experiencing the same login issues.

**Reply by Firstyear** [reply]
What was the error you saw? Once it has posix attrs it should work.

**Reply by Firstyear** [reply]
Also UPG as the allowed login group should also have worked.

**Reply by Firstyear** [reply]
does the group resolve on the machine? `getent group usuario1`

**Reply by frateralexander** [reply]
Hi, thanks for your response!

To answer the questions you raised:

When I use "getent" to check the user and group status, everything seems normal:

```
$ getent passwd usuario1
usuario1@localhost:x:1004:1004:User 1:/home/usuario1@localhost:/bin/bash
```
```
$ getent group usuario1
usuario1@localhost:x:1004:usuario1@localhost
```
Other tests on the user and group also look correct:

```
$ kanidm group posix show usuario1
[...]
[ spn: usuario1@localhost, gidnumber: 1004, name: usuario1, uuid: a2423599-926a-4b2d-b3af-0ff9eb3d31f6 ]
```
```
$ kanidm person posix show usuario1
[...]
---
spn: usuario1@localhost
name: usuario1
displayname: User 1
uuid: a2423599-926a-4b2d-b3af-0ff9eb3d31f6
shell: /bin/bash
group: [ spn: usuario1@localhost, gidnumber: 1004, name: usuario1, uuid: a2423599-926a-4b2d-b3af-0ff9eb3d31f6 ]
```
However, when I try to use su - usuario1, I still get the following error:

```
$ su - usuario1
su: Authentication failure
```

And the logs show this:

`Cannot authenticate users, no allowed groups in configuration!
`

Here is the configuration I’m using:

/etc/kanidm/config:

```
uri = "https://localhost:443"
verify_ca = false
```

/etc/kanidm/unixd:

```
version = '2'
[kanidm]
pam_allowed_login_groups = ["idm_all_accounts", "posix_group"]
```
To create the user, I used the following commands:

```
kanidm person create usuario1 "User 1"
kanidm person posix set --gidnumber 1004 --shell /bin/bash usuario1
kanidm person posix set-password usuario1
```

Everything seems to resolve fine, but the login still fails. Any insights on what might be missing in my setup?

Thanks for your help!

**Reply by Firstyear** [reply]
If you do `id usurario1` what groups does it show?

**Reply by Firstyear** [reply]
You also need to check `getent group posix_group` and `getent group idm_all_accounts`

**Reply by frateralexander** [reply]
Hi, thanks for the follow-up!

Here are the results of the checks you requested:

Output of id usuario1:

```
$ id usuario1
uid=1004(usuario1@localhost) gid=1004(usuario1@localhost) groups=1004(usuario1@localhost)
```

Checking the group idm_all_accounts:

```
$ getent group idm_all_accounts
**(no output)**
```

I will not be using the posix_group anyway , so this test is focused solely on idm_all_accounts. Using kanidm to verify idm_all_accounts, I see that usuario1@localhost is listed as a dynamic member:

```
$ kanidm group get idm_all_accounts
[...]
dynmember: usuario1@localhost
[...]
```
While idm_all_accounts seems to be resolving correctly via the kanidm command, the getent queries for idm_all_accounts return no output, which may be part of the issue.

Do you have any suggestions on what might be causing getent not to resolve this group, even though usuario1 appears to be a valid member of idm_all_accounts?

Thanks again for your assistance!

**Reply by Firstyear** [reply]
The fact that "posix_group" and "idm_all_accounts" are not listed in `getent group` means that they aren't *posix groups* and that's why they don't work to control access with pam_allowed_login_groups. 

https://kanidm.github.io/kanidm/stable/accounts/posix_accounts_and_groups.html#enabling-posix-attributes-on-groups
